compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -1972,12 +1972,30 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
                 // Pick the first substitution that still contains inference variables as the one
                 // we're going to emit an error for. If there are none (see above), fall back to
                 // the substitution for `Self`.
-                let subst = {
-                    let substs = data.trait_ref.substs;
-                    substs
-                        .iter()
-                        .find(|s| s.has_infer_types_or_consts())
-                        .unwrap_or_else(|| substs[0])
+                let subst = data.trait_ref.substs.iter().find(|s| s.has_infer_types_or_consts());
+
+                let mut err = if let Some(subst) = subst {
+                    let impl_candidates = self
+                        .find_similar_impl_candidates(trait_ref)
+                        .into_iter()
+                        .map(|candidate| candidate.trait_ref)
+                        .collect();
+                    self.emit_inference_failure_err(
+                        body_id,
+                        span,
+                        subst,
+                        impl_candidates,
+                        ErrorCode::E0283,
```

Comment:
```suggestion
                    let error_code = if self.tcx.lang_items().sized_trait() == Some(trait_ref.def_id()) { if self.is_tainted_by_errors() { return } else { E0282 } } else { E0283 };
                    self.emit_inference_failure_err(
                        body_id,
                        span,
                        subst,
                        impl_candidates,
                        error_code,
```
something like this and removing the branching below should be cleaner (with the right formatting ofc)

Reply:
Unifying the control flow here and below is actually problematic. The early `emit` + `return` here in the case of unsatisfied `Sized` predicates is important -- for sized errors, we don't want to do any of the annotations below, like:
* "cannot satisfy `_: Sized`"
* "consider specifying the type arguments in the function call"

Reply:
Also we don't really want to point out things like this:

```
LL | pub enum Result<T, E> {
   |                    ^ required by this bound in `Ok`
``` 

when the `E` parameter is just ambiguous.

Reply:
hmm, can you then move the `Sized` check above this? Prefer to not cancel diagnostics unless necessary?

Reply:
Can do.